Rework single-detector stat

spxiwh opened the pull request:
At the moment if one wants to try a new detection statistic they edit the `stat` module, but then they have to edit a bunch of different codes to allow that statistic to be used in the single-detector applications (minifollowups, plots, the fitting codes, ...). This can be a little frustrating.

As such I want to move all of this into the stat module. Of course the single-detector statistics are different from the multi-detector ones, but in most cases calling the "single" method of the Stat class gets what we want.

This patch implements the functionality where a "single-detector-statistic" dict is defined in the stat module, and this is inherited everywhere else. We might envisage Stat classes that might *only* have single methods, and only be valid for single-detector statistics, but for now we are just using existing classes to implement what we want.

spxiwh commented:
This has been tested as part of the IMBH + ML project, I'm also going to ask Collin to test this.

spxiwh commented:
@tdent Poke on this (I don't believe the Codacy issue is a problem, we could make the dictionary "unmutable", but it probably doesn't help anything).

tdent commented:
Hi Ian, started looking at this. The approach of accessing sngl statistic values via some coinc stat class that uses them seems a bit convoluted to me .. why is this the preferred approach for 'common' single detector stats that one might want to access directly?

spxiwh commented:
Hi @tdent, so the main thing I want to solve here is that the list of single-detector statistics is hard-coded, so I want to move that to the stat module as a "one place to look for that information".

I figured that as the various classes in that module provide a "single" method, that would be a good thing to use to access this information. While that might seem overkill for things like "newsnr", which are simple functions, some things like the "ExpFitCombinedSNR" class use a bunch of things defined at the class level, and it didn't make sense to duplicate that.  It is true that for some classes the "single" function might not return anything useful, as the statistic is not defined at all without a coincident trigger. I'm open to other suggestions for doing this though, but feel strongly that having codes look into the stat module for this information is the right way to proceed.

tdent commented:
ok, I agree that it should be possible to access the 'single' stats from the coinc stat classes, along with generic single-ifo stats, and stat.py is a reasonable place for doing that.  I was having some difficulty in following the route of doing so, though (dictionaries / instances / .. ?)

spxiwh commented:
Hi @tdent, can we consider this pull request again now?

tdent commented:
I have been able to consider it all the time, but I'm afraid starting at a new institution and trying to organize everything that is and could be needed for the impending papers has had *much* higher priority.

tdent commented on bin/hdfcoinc/pycbc_fit_sngls_binned:
```diff
@@ -27,27 +27,14 @@ from scipy.stats import kstest
-    else:
-        return stat_dict[statchoice](snr, rchisq)
+def get_stat(statchoice, trigs):
+    stat_instance = sngl_statistic_dict[statchoice]([])
```
add comment to this line e.g. 'initialize the stat class with an empty file list - in general could also feed it precalculated files'

tdent commented on pycbc/events/stat.py:
```diff
@@ -44,8 +44,8 @@ def get_newsnr(trigs):
     """
-    dof = 2. * trigs['chisq_dof'] - 2.
-    newsnr = events.newsnr(trigs['snr'], trigs['chisq'] / dof)
+    dof = 2. * trigs['chisq_dof'][:] - 2.
```
so here, trigs can be either a dictionary of ndarrays but also a h5py *group* which then yields *datasets* which then have to be indexed on [:] to give numpy arrays.  hence will ask for one tweaking of the documentation ..

tdent replied:
i.e. in the description of 'trigs' above add that it can also be an h5py group with datasets analogous to the required numpy arrays

tdent commented:
OK, this looks reasonable .. please fix the description of 'trigs' in the statistic 'single()' methods to not just specify dicts but also h5py groups. 

Not sure quite what to do about the Codacy thing .. would it not work to replace it by two functions, get_coinc_statistic and get_sngl_statistic, each using its own (hard-coded) dictionary?  Not very elegant but actually one less level of complexity (i.e. one less function call if using get_sngl_statistic).

tdent commented:
also should we add Alex as a reviewer?

spxiwh commented:
Thanks @tdent I'll fix the issues you raise and bring the patch back up to date.

I think the fix for the Codacy issue is to make the object a *static* dictionary (same as the tuple/list distinction) so that it couldn't be changed within the function (although there's no danger in this case as the function does not edit the dictionary).

Also if @ahnitz wants to comment on this he should, I'm sure he's seen it.

spxiwh commented:
Turns out `frozendict` was rejected to be added to python:

https://www.python.org/dev/peps/pep-0416/

There is a python3 "method" for dealing with this case, so I'll use that when we want to drop python2 support. For now I'll ignore the codacy warning, the function *is* safe in this regard

spxiwh commented on pycbc/events/stat.py:
```diff
@@ -523,14 +525,24 @@ def single(self, trigs):
+    'newsnr_sgveto': NewSNRSGStatistic
+}
+
+def get_statistic(stat, stat_dict=statistic_dict):
```
Yeah, I also originally wrote it in that way, but CodeClimate asked me to change it. I reverted that change now.

spxiwh commented:
Okay, I think I've addressed the comments (and it's a good thing that we have an admin lock on origin/master, as otherwise I would have done something very bad!)

tdent commented on pycbc/events/stat.py:
```diff
@@ -539,14 +542,24 @@ def single(self, trigs):
+    'newsnr_sgveto': NewSNRSGStatistic
+}
+
 def get_statistic(stat):
```
might consider renaming this to get_coinc_statistic() .. that would clearly require a bunch of other changes though

spxiwh replied:
Yeah, I did consider that, but decided it wasn't worth the potential for breaking other code by renaming this. I can do this if you think its necessary though.

tdent replied:
Another day.

spxiwh commented:
> I don't think anything here is _very_ bad .. unless it's something else I'm not seeing

Yeah, that's because the admin lock said "Action forbidden" and stopped me accidentally "push -f"ing to origin/master
